### PR TITLE
Redesigning RideHail AV charging in BEAM #WIP

### PR DESCRIPTION
**Pull request by haitamlaarabi**
DONE:
+ Currently RH AV are able to charge via the CNM
TODO:
+ Test the both AV and Human Ridehail charging
+ Create a test for AV and Human Ridehail charging
+ Create a test for queueing at depots (only Ridehail AV)
+ Test a production run

- - -
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594)

### CODE REVIEW

**Comment by nikolayilyin** [comment]
test!

**Review by dimaopen — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594)* status: 0 of 44 files reviewed, 6 unresolved discussions (waiting on @haitamlaarabi, @heyrutvik, @nikolayilyin, and @wrashid)

___
*[`src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala` line 460 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5G6qzECwxr6AApqDl:-NA5G6qzECwxr6AApqDm:b4bpneq) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala#L460)):*
> ```Scala
>           }
>       })
>       .map { result =>
> ```

we could user `foreach` instead of `map`.
___
*[`src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala` line 472 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5FDHnBse-V118ZgDo:-NA5FDHnBse-V118ZgDp:b-sd6ixa) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala#L472)):*
> ```Scala
>     * @return
>     */
>   def sendChargingInquiry(whenWhere: SpaceTime, beamVehicle: BeamVehicle, triggerId: Long): Future[Any] = {
> ```

I'd suggest return `Future[ParkingInquiryResponse]` here for type safety and clarity.



_Code quote:_
```Scala
sendChargingInquiry
```
___
*[`src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala` line 832 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5IXr43K6QpNSYEN6A:-NA5IXr43K6QpNSYEN6B:b-ww0s5c) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala#L832)):*
> ```Scala
>           additionalCustomVehiclesForDepotCharging,
>           triggerId
>         ) =>
> ```

This case branch is very big. It is hard to follow the cases. Is it possible to extract this case branch into a method?
___
*[`src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala` line 921 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5Hsxt4HksYN5Xka66:-NA5Hsxt4HksYN5Xka67:bys3r3f) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala#L921)):*
> ```Scala
> 
>       if (repositionVehicles.isEmpty) {
>         log.debug("sendCompletionAndScheduleNewTimeout from 1486")
> ```

What does 1486 mean?
___
*[`src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala` line 929 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5JMPPCT6UV4aTdnmS:-NA5JMPPCT6UV4aTdnmT:bvtdue1) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala#L929)):*
> ```Scala
>       }
> 
>       val futureRepoRoutingMap = mutable.Map[Id[BeamVehicle], Future[RoutingRequest]]()
> ```

Could we use a collection `map` method instead of adding values to a mutable structure? Standard `map` method means that every item from one collection is transformed and put into the new collection. If we use a mutable structure then we need to carefully look through the code to understand it.
___
*[`src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala` line 933 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5K-0q34uRpmLvjJS3:-NA5K-0q34uRpmLvjJS4:b-swavw8) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala#L933)):*
> ```Scala
>       for ((vehicleId, destinationLocation) <- repositionVehicles) {
>         rideHailManagerHelper.getServiceStatusOf(vehicleId) match {
>           case _ =>
> ```

Why do we use pattern matching with "everything" pattern?

**Review by haitamlaarabi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594)* status: 0 of 44 files reviewed, 6 unresolved discussions (waiting on @dimaopen, @haitamlaarabi, and @JustinPihony)

___
*[`src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala` line 460 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5G6qzECwxr6AApqDl:-NA7-kGu91gRYxxysyUp:b-896fix) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala#L460)):*
<details><summary><i>Previously, dimaopen (Dmitrii Openkov) wrote…</i></summary><blockquote>

we could user `foreach` instead of `map`.
</blockquote></details>

Done.
___
*[`src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala` line 472 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5FDHnBse-V118ZgDo:-NA7-mFnAm996Qdq-iKy:b-896fix) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala#L472)):*
<details><summary><i>Previously, dimaopen (Dmitrii Openkov) wrote…</i></summary><blockquote>

I'd suggest return `Future[ParkingInquiryResponse]` here for type safety and clarity.
</blockquote></details>

Done.
___
*[`src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala` line 832 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5IXr43K6QpNSYEN6A:-NA705YqE9HjZ85BZd2I:b-clwexr) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala#L832)):*
<details><summary><i>Previously, dimaopen (Dmitrii Openkov) wrote…</i></summary><blockquote>

This case branch is very big. It is hard to follow the cases. Is it possible to extract this case branch into a method?
</blockquote></details>

It was under a different case. I just put that entire case inside a method



_Code quote:_
```Scala
case ParkingStallsClaimedByVehicles
```
___
*[`src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala` line 921 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5Hsxt4HksYN5Xka66:-NA7151Y5DqqNdNgeeFT:bgqpaa0) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala#L921)):*
<details><summary><i>Previously, dimaopen (Dmitrii Openkov) wrote…</i></summary><blockquote>

What does 1486 mean?
</blockquote></details>

No idea. Maybe referring to a line. This is an old code.



_Code quote:_
```Scala
sendCompletionAndScheduleNewTimeout from 1486
```
___
*[`src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala` line 933 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5K-0q34uRpmLvjJS3:-NA72QnA4PpD_TBHnD9w:b41lc39) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala#L933)):*
<details><summary><i>Previously, dimaopen (Dmitrii Openkov) wrote…</i></summary><blockquote>

Why do we use pattern matching with "everything" pattern?
</blockquote></details>

good question but it's an old code. I'm deleting it



_Code quote:_
```Scala
rideHailManagerHelper.getServiceStatusOf(vehicleId) match {
```

**Comment by haitamlaarabi** [comment]
Test now!

**Comment by haitamlaarabi** [comment]
Test now!

**Comment by haitamlaarabi** [comment]
Test now!

**Comment by haitamlaarabi** [comment]
Test now!

File: src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala [hunk 1]
```diff
@@ -235,66 +346,87 @@ class DefaultRideHailDepotParkingManager(
...
+            )
+          case _ => false
+        }
     }
+    Future
+      .sequence(idleVehicleIdsWantingToRefuelWithLocation.map { case (vehicleId, rideHailAgentLocation) =>
+        val beamVehicle = resources(vehicleId)
+        val locationUtm: Location = rideHailAgentLocation.getCurrentLocationUTM(tick, beamServices)
+        sendChargingInquiry(SpaceTime(locationUtm, tick), beamVehicle, triggerId)
+          .mapTo[ParkingInquiryResponse]
```
Comment (dimaopen) [inline src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala]: This `mapTo` could be removed.

**Review by dimaopen — APPROVED** [review]
Approved with a minor note

**Review by haitamlaarabi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594)* status: 0 of 48 files reviewed, 7 unresolved discussions (waiting on @dimaopen and @JustinPihony)

___
*[`src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala` line 384 at r10](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-N9soVaBD4rN2hmo3lBo-r10-384:-NABp9VyCmbuGNpTN0hL:b-896fix) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/46f599b37872344e8f65ef1a65619e66eed1bccb/src/main/scala/beam/agentsim/agents/ridehail/DefaultRideHailDepotParkingManager.scala#L384)):*
<details><summary><i>Previously, dimaopen (Dmitrii Openkov) wrote…</i></summary><blockquote>

This `mapTo` could be removed.
</blockquote></details>

Done.



_Code quote:_
```Scala
        sendChargingInquiry(SpaceTime(locationUtm, tick), beamVehicle, triggerId)
          .mapTo[ParkingInquiryResponse]
```
___
*[`src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala` line 929 at r6](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3594#-NA5JMPPCT6UV4aTdnmS:-NABpYe0BKxXbNCRwAP1:b-896fix) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/e578aabb2119322bb474a8bca6710ce3004ae6f1/src/main/scala/beam/agentsim/agents/ridehail/RideHailManager.scala#L929)):*
<details><summary><i>Previously, dimaopen (Dmitrii Openkov) wrote…</i></summary><blockquote>

Could we use a collection `map` method instead of adding values to a mutable structure? Standard `map` method means that every item from one collection is transformed and put into the new collection. If we use a mutable structure then we need to carefully look through the code to understand it.
</blockquote></details>

Done.



_Code quote:_
```Scala
 var futureRepoRoutingMap = Map[Id[BeamVehicle], Future[RoutingRequest]]()
```

**Comment by haitamlaarabi** [comment]
Test now!

**Comment by haitamlaarabi** [comment]
Test now!

**Comment by haitamlaarabi** [comment]
Test now!

**Comment by haitamlaarabi** [comment]
Test now!

File: src/main/scala/beam/agentsim/Resource.scala [hunk 2]
```diff
@@ -7,15 +7,17 @@ import beam.agentsim.agents.vehicles.BeamVehicle.BeamVehicleState
 import beam.agentsim.agents.vehicles.{BeamVehicle, PassengerSchedule}
 import beam.agentsim.events.SpaceTime
 import beam.agentsim.scheduler.HasTriggerId
+import beam.agentsim.scheduler.Trigger
 import beam.sim.Geofence
 import org.matsim.api.core.v01.Id
 
 object Resource {
 
-  case class ReleaseParkingStall(stall: ParkingStall, triggerId: Long) extends HasTriggerId
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/Resource.scala:15]: Should we drop the triggerId? We can use TriggerWithId maybe?

File: src/main/scala/beam/agentsim/agents/PersonAgent.scala [hunk 3]
```diff
@@ -1593,22 +1577,21 @@ class PersonAgent(
...
+      log.error("myUnhandled.UnhandledVehicle: {}", ev)
       stay()
-    case ev @ Event(WaitingToCharge(_, _, _), _) =>
+    case ev @ Event(_: WaitingToCharge, _) =>
       log.debug("myUnhandled.WaitingInLine: {}", ev)
       stay()
-    case ev @ Event(EndingRefuelSession(_, _, triggerId), _) =>
+    case ev @ Event(_: EndingRefuelSession, _) =>
       log.debug("myUnhandled.EndingRefuelSession: {}", ev)
-      scheduler ! CompletionNotice(triggerId)
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/PersonAgent.scala:1611]: In case this was an accident I wanted to point it out as it could lead to stuckness.

File: src/main/scala/beam/sim/BeamMobsim.scala [hunk 4]
```diff
@@ -204,7 +204,7 @@ class BeamMobsim @Inject() (
         case VehicleCategory.Bike => BeamMode.BIKE
       }.toList
 
-      val cavs = vehicles.filter(_.beamVehicleType.isCav).toList
+      val cavs = vehicles.filter(_.isCAV).toList
```
Comment (JustinPihony) [inline src/main/scala/beam/sim/BeamMobsim.scala:207]: Should we rename all cav to connectedautomationvehicle as used elsewhere?

File: src/main/scala/beam/sim/RideHailFleetInitializer.scala [hunk 5]
```diff
@@ -319,7 +319,7 @@ object RideHailFleetInitializer extends OutputDataDescriptor with LazyLogging {
       val powertrain = new Powertrain(beamVehicleType.primaryFuelConsumptionInJoulePerMeter)
 
       val managerIdDependsOnWhetherVehicleIsCav =
-        if (beamVehicleType.isCav) rideHailManagerId else VehicleManager.AnyManager.managerId
+        if (beamVehicleType.isFullSelfDriving) rideHailManagerId else VehicleManager.AnyManager.managerId
```
Comment (JustinPihony) [inline src/main/scala/beam/sim/RideHailFleetInitializer.scala]: Should this be cav || fullselfdriving?

File: src/main/scala/beam/agentsim/infrastructure/ParkingNetworkManager.scala [hunk 6]
```diff
@@ -43,7 +48,42 @@ class ParkingNetworkManager(beamServices: BeamServices, parkingNetworkMap: Parki
 
 object ParkingNetworkManager extends LazyLogging {
 
+  val rnd = new Random()
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/infrastructure/ParkingNetworkManager.scala]: This appears to not be used.

**Comment by haitamlaarabi** [comment]
Test now!

**Comment by haitamlaarabi** [comment]
Test now!

File: src/main/scala/beam/agentsim/Resource.scala (hunk 2 above)
Comment (haitamlaarabi) [reply src/main/scala/beam/agentsim/Resource.scala:15]: We never use the triggerId when we releaseParkingStall. I prefer simplifying the code to not assume that a certain trigger is being completed through a ReleaseParking.

File: src/main/scala/beam/agentsim/agents/PersonAgent.scala (hunk 3 above)
Comment (haitamlaarabi) [reply src/main/scala/beam/agentsim/agents/PersonAgent.scala:1611]: We are completing that trigger in the ChargingNetworkManager

File: src/main/scala/beam/sim/BeamMobsim.scala (hunk 4 above)
Comment (haitamlaarabi) [reply src/main/scala/beam/sim/BeamMobsim.scala:207]: Whoever does not understand what CAV means, can see the meaning here: 
`  def isCAV: Boolean = beamVehicleType.isConnectedAutomatedVehicle`
There is not need to replace it elsewhere

File: src/main/scala/beam/sim/RideHailFleetInitializer.scala (hunk 5 above)
Comment (haitamlaarabi) [reply src/main/scala/beam/sim/RideHailFleetInitializer.scala]: I have resolved this confusing. As of now, all CAVs have a level of automation >= 4 and are considered by BEAM as robot taxis. No need to separate between FullSelfDriving (level 5) and level 4

File: src/main/scala/beam/agentsim/infrastructure/ParkingNetworkManager.scala (hunk 6 above)
Comment (haitamlaarabi) [reply src/main/scala/beam/agentsim/infrastructure/ParkingNetworkManager.scala]: True

**Comment by haitamlaarabi** [comment]
Test now!

**Comment by haitamlaarabi** [comment]
Test now!
